addons/crm_mail_client_extension/controllers/main.py

```diff
@@ -6,31 +6,111 @@
+        # are connected to the right company before performing such operation.
+        record = request.env['crm.lead'].create({
+            'name': html2plaintext(email_subject),
+            'partner_id': partner_id,
+            'description': html2plaintext(email_body),
+        })
+
+        return {'lead_id': record.id}
+
+    @http.route('/mail_client_extension/lead/open', type='http', auth='user')
+    def crm_lead_open(self, lead_id):
+        action = http.request.env.ref("crm.crm_lead_view_form")
+        url = '/web#id=%s&action=%s&model=crm.lead&edit=1&model=crm.lead' % (lead_id, action.id)
+        return werkzeug.utils.redirect(url)
+
+    def _fetch_partner_leads(self, partner, limit=5, offset=0):
+        """
+        Returns an array containing partner leads, each lead will have the following structure :
+        {
+            id: the lead's id,
+            name: the lead's name,
+            expected_revenue: the expected revenue field value
+            probability: the value of the probability field,
+            recurring_revenue: the value of the recurring_revenue field if the lead has a recurring revenue
+            recurring_plan: the value of the recurring plan field if the lead has a recurring revenue
+        }
+        """
+
+        partner_leads = request.env['crm.lead'].search(
+            [('partner_id', '=', partner.id)], offset=offset, limit=limit)
+        recurring_revenues = request.env.user.has_group('crm.group_use_recurring_revenues')
+
         leads = []
         for lead in partner_leads:
-            leads.append({
+            lead_values = {
                 'id': lead.id,
                 'name': lead.name,
-                'expected_revenue': formatLang(request.env, lead.expected_revenue, monetary=True, currency_obj=lead.company_currency),
-            })
+                'expected_revenue': formatLang(request.env, lead.expected_revenue, monetary=True,
+                                               currency_obj=lead.company_currency),
+                'probability': lead.probability,
+            }
 
-        return {'leads': leads}
+            if recurring_revenues:
+                lead_values.update({
+                    'recurring_revenue': formatLang(request.env, lead.recurring_revenue, monetary=True,
+                                                    currency_obj=lead.company_currency),
+                    'recurring_plan': lead.recurring_plan.name,
+                })
 
-    @http.route('/mail_client_extension/lead/create_from_partner', type='http', auth='user', methods=['GET'])
-    def crm_lead_redirect_form_view(self, partner_id):
-        server_action = http.request.env.ref("crm_mail_client_extension.lead_creation_prefilled_action")
-        return werkzeug.utils.redirect('/web#action=%s&model=crm.lead&partner_id=%s' % (server_action.id, int(partner_id)))
+            leads.append(lead_values)
+
+        return leads
+
+    def _prepare_contact_values(self, partner):
```

Comment:
We could probably split mail_client_extension overrides from crm-specific routes (like already proposed on helpdesk).

Reply:
Done,

The controller code has been split between the "mail_plugin.py" and "crm_client.py" files